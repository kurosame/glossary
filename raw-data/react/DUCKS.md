## category

react

## titles

Ducks
Redux

## description

Redux のデザインパターン  
Action や Reducer を modules にパッケージする

vuejs-boilerplate でも使ってる気がする  
1 ファイルにはしてないけど、actions, getters, mutations, state を import して modules でパッケージして export してる  
そして Store 作るときに modules をセットしている
