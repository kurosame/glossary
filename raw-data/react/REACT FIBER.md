## category

react

## titles

React Fiber
react-fiber

## description

Begin, Complete, Commit のフェーズがある

### Begin, Complete

Component 単位で実行される  
componentWillMount, componentWillReceiveProps, shouldComponentUpdate, componentWillUpdate  
非同期レンダリングでは中断・再開により、複数回ライフサイクル関数が呼ばれる可能性がある

### Commit

全ての Fiber が完了した時に実行される  
DOM 更新などの副作用が発生する処理はここでまとめて反映される  
componentDidMount, componentDidUpdate, componentWillUnmount

### Time Slicing

Fiber により画面のレンダリングが非同期になると、Fiber 単位で処理の優先度を付けることができる  
例えばユーザ操作の優先度を高くすることで、画面がレンダリング中でもユーザ操作を優先して処理することが可能となる

### Suspense

非同期通信をしている間、画面のレンダリングを中断できる  
非同期通信が完了した後に、画面をレンダリングすることが可能  
render に Promise を投げて、その Promise が完了した後に render するみたいな？
