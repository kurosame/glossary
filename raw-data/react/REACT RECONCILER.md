## category

react

## titles

React Reconciler
react-reconciler

## description

VDOM と呼ばれている部分

1. ReactElement のツリーを構築
1. Component をマウント、アンマウント、更新
1. ライフサイクル関数を呼出す
1. 現在の DOM との差分を計算する
1. renderer に差分を伝える
