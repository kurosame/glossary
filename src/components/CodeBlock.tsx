import React from 'react'

import CodeHighlighter from '@/components/CodeHighlighter'

interface Props {
  inline?: boolean
  className?: string
  children: React.ReactNode & React.ReactNode[]
}

const CodeBlock: React.VFC<Props> = ({ inline, className, children, ...props }) => {
  const match = /language-(\w+)/.exec(className || '')
  return !inline && match ? (
    <CodeHighlighter language={match[1]}>{children}</CodeHighlighter>
  ) : (
    <code className={className} {...props}>
      {children}
    </code>
  )
}

export default CodeBlock
