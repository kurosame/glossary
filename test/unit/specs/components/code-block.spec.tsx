/**
 * @jest-environment jsdom
 */
import React from 'react'

import { cleanup, render, RenderResult } from '@testing-library/react'

import CodeBlock from '@/components/CodeBlock'

let wrapper: (inline?: boolean, className?: string) => RenderResult
beforeEach(() => {
  wrapper = (inline, className): RenderResult =>
    render(
      <CodeBlock inline={!!inline} className={className ?? ''}>
        {['']}
      </CodeBlock>
    )
})
afterEach(cleanup)

describe('Render to `CodeHighlighter`', () => {
  test('Returns `CodeHighlighter` when matches a regular expression', () => {
    expect(wrapper(undefined, 'language-js').queryByTestId('code-highlighter')).not.toBeNull()
  })

  test('Match the snapshot', () => {
    expect(wrapper(undefined, 'language-js').asFragment()).toMatchSnapshot()
  })
})

describe('Render to `<code></code>`', () => {
  test('Returns `<code></code>` when not matches a regular expression', () => {
    expect(wrapper().queryByTestId('code-highlighter')).toBeNull()
  })

  test('Match the snapshot', () => {
    expect(wrapper().asFragment()).toMatchSnapshot()
  })
})
