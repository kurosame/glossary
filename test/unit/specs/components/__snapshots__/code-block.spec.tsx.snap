// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render to \`<code></code>\` Match the snapshot 1`] = `
<DocumentFragment>
  <code />
</DocumentFragment>
`;

exports[`Render to \`CodeHighlighter\` Match the snapshot 1`] = `
<DocumentFragment>
  <pre
    data-testid="code-highlighter"
    style="color: rgb(197, 200, 198); text-shadow: 0 1px rgba(0, 0, 0, 0.3); font-family: Inconsolata, Monaco, Consolas, 'Courier New', Courier, monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 4; hyphens: none; padding: 1em; margin: .5em 0px; overflow: auto; border-radius: 0.3em; background: rgb(29, 31, 33);"
  >
    <code
      class="language-js"
      style="color: rgb(197, 200, 198); text-shadow: 0 1px rgba(0, 0, 0, 0.3); font-family: Inconsolata, Monaco, Consolas, 'Courier New', Courier, monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 4; hyphens: none;"
    />
  </pre>
</DocumentFragment>
`;
